pkg/controller/account/iam.go

```diff
@@ -0,0 +1,185 @@
+
+// Type for JSON response from Federation end point
+type awsSigninTokenRepsonse struct {
+	SigninToken string
+}
+
+// RequestSigninToken makes a HTTP request to retrieve a Signin Token from the federation end point
+func RequestSigninToken(reqLogger logr.Logger, awsclient awsclient.Client, DurationSeconds *int64, FederatedUserName *string, PolicyArns []*sts.PolicyDescriptorType, STSCredentials *sts.AssumeRoleOutput) (string, error) {
+
+	// // URLs for building Federated Signin queries
+	federationEndPointURL := "https://signin.aws.amazon.com/federation"
+	awsConsoleURL := "https://console.aws.amazon.com/"
+	issuer := "Red Hat SRE"
+
+	GetFederationTokenInput := sts.GetFederationTokenInput{
+		DurationSeconds: DurationSeconds,
+		Name:            FederatedUserName,
+		PolicyArns:      PolicyArns,
+	}
+
+	// Get Federated token credentials to build console URL
+	GetFederationTokenOutput, err := awsclient.GetFederationToken(&GetFederationTokenInput)
+
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			// Get error details
+			reqLogger.Error(err, fmt.Sprintf("Error: %s, %s", awsErr.Code(), awsErr.Message()))
+			return "", err
+		}
+
+		return "", err
+	}
+	// Build JSON credentials for federation requets
+	federationCredentials := map[string]string{
+		"sessionId":    *GetFederationTokenOutput.Credentials.AccessKeyId,
+		"sessionKey":   *GetFederationTokenOutput.Credentials.SecretAccessKey,
+		"sessionToken": *GetFederationTokenOutput.Credentials.SessionToken,
+	}
+
+	jsonCredentials, err := json.Marshal(federationCredentials)
+
+	// TODO better error here
+	if err != nil {
+		fmt.Println("Error serializing json")
+		fmt.Println(err)
+		return "", err
+	}
+
+	// Build URL to request Signin Token via Federation end point
+	baseFederationURL, err := url.Parse(federationEndPointURL)
+
+	if err != nil {
+		fmt.Println("Malformed URL: ", err.Error())
+		return "", err
+	}
+
+	federationParams := url.Values{}
+
+	federationParams.Add("Action", "getSigninToken")
+	federationParams.Add("SessionType", "json")
+	federationParams.Add("Session", string(jsonCredentials))
+
+	baseFederationURL.RawQuery = federationParams.Encode()
+
+	// Make HTTP request to retrieve Federated Signin Token
+	res, err := http.Get(baseFederationURL.String())
+
+	if err != nil {
+		fmt.Printf("Error requesting Signin Token from: %s\n", baseFederationURL.String())
+		return "", err
+	}
+
+	defer res.Body.Close()
+
+	body, err := ioutil.ReadAll(res.Body)
+
+	if err != nil {
+		fmt.Printf("Unable to read response body: %s", baseFederationURL.String())
+		return "", err
+	}
+
+	var SigninResponse awsSigninTokenRepsonse
+
+	// Unmarshal JSON response so we can extract the signin token
+	err = json.Unmarshal(body, &SigninResponse)
+
+	if err != nil {
+		fmt.Printf("Error unmarshalling Federated Signin Response JSON")
+		return "", err
+	}
+
+	signinFederationURL, err := url.Parse(federationEndPointURL)
+	if err != nil {
+		fmt.Println("Malformed URL: ", err.Error())
```

Comment:
Are these debug errors or should this be a log

Reply:
Should there be a return or error handling here ?

Reply:
Fixed